heat/core/manipulations.py

```diff
@@ -1,18 +1,275 @@
 ]
 
 
+def concatenate(arrays, axis=0):
+    """
+    Join 2 arrays along an existing axis.
+
+    Parameters
+    ----------
+    arrays: tuple of 2 DNDarrays
+        The arrays must have the same shape, except in the dimension corresponding to axis (the first, by default).
+    axis: int, optional
+        The axis along which the arrays will be joined. Default is 0.
+
+    Returns
+    -------
+    res: DNDarray
+        The concatenated DNDarray
+
+    Examples
+    --------
+    >>> x = ht.zeros((3, 5), split=None)
+    [0/1] tensor([[0., 0., 0., 0., 0.],
+    [0/1]         [0., 0., 0., 0., 0.],
+    [0/1]         [0., 0., 0., 0., 0.]])
+    [1/1] tensor([[0., 0., 0., 0., 0.],
+    [1/1]         [0., 0., 0., 0., 0.],
+    [1/1]         [0., 0., 0., 0., 0.]])
+    >>> y = ht.ones((3, 6), split=0)
+    [0/1] tensor([[1., 1., 1., 1., 1., 1.],
+    [0/1]         [1., 1., 1., 1., 1., 1.]])
+    [1/1] tensor([[1., 1., 1., 1., 1., 1.]])
+    >>> ht.concatenate((x, y), axis=1)
+    [0/1] tensor([[0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 1.],
+    [0/1]         [0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 1.]])
+    [1/1] tensor([[0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 1.]])
+
+    >>> x = ht.zeros((4, 5), split=1)
+    [0/1] tensor([[0., 0., 0.],
+    [0/1]         [0., 0., 0.],
+    [0/1]         [0., 0., 0.],
+    [0/1]         [0., 0., 0.]])
+    [1/1] tensor([[0., 0.],
+    [1/1]         [0., 0.],
+    [1/1]         [0., 0.],
+    [1/1]         [0., 0.]])
+    >>> y = ht.ones((3, 5), split=1)
+    [0/1] tensor([[1., 1., 1.],
+    [0/1]         [1., 1., 1.],
+    [0/1]         [1., 1., 1.]])
+    [1/1] tensor([[1., 1.],
+    [1/1]         [1., 1.],
+    [1/1]         [1., 1.]])
+    >>> ht.concatenate((x, y), axis=0)
+    [0/1] tensor([[0., 0., 0.],
+    [0/1]         [0., 0., 0.],
+    [0/1]         [0., 0., 0.],
+    [0/1]         [0., 0., 0.],
+    [0/1]         [1., 1., 1.],
+    [0/1]         [1., 1., 1.],
+    [0/1]         [1., 1., 1.]])
+    [1/1] tensor([[0., 0.],
+    [1/1]         [0., 0.],
+    [1/1]         [0., 0.],
+    [1/1]         [0., 0.],
+    [1/1]         [1., 1.],
+    [1/1]         [1., 1.],
+    [1/1]         [1., 1.]])
+    """
+
+    if len(arrays) < 2:
+        raise ValueError('concatenate requires 2 arrays')
+    elif len(arrays) > 2:
+        res = concatenate((arrays[0], arrays[1]), axis=axis)
+        for a in range(2, len(arrays)):
+            res = concatenate((res, arrays[a]), axis=axis)
+        return res
+
+    arr0, arr1 = arrays[0], arrays[1]
+    out_dtype = types.promote_types(arr0.dtype, arr1.dtype)
+    if arr0.dtype != out_dtype:
+        arr0 = out_dtype(arr0)
+    if arr1.dtype != out_dtype:
+        arr1 = out_dtype(arr1)
+
+    if not isinstance(arr0, dndarray.DNDarray) and not isinstance(arr1, dndarray.DNDarray):
+        raise TypeError('Both arrays must be DNDarrays')
+    if not isinstance(axis, int):
+        raise TypeError('axis must be an integer, currently: {}'.format(type(axis)))
```

Comment:
I suppose you have a good reason not to call sanitize_axis? At the moment ht.concatenate fails if axis<0. 

Example axis>0, this one works fine:

```
a = ht.array(ht.random.randn(2, 4), split=1)
b = ht.array(ht.random.randn(2, 7), split=1)
c = ht.concatenate((a, b), axis=1)
```

Equivalent with axis<0, doesn't run, see below:

```
a = ht.array(ht.random.randn(2, 4), split=1)
b = ht.array(ht.random.randn(2, 7), split=1)
c = ht.concatenate((a, b), axis=-1)
```

> > mpirun --use-hwthread-cpus -n 2 -tag-output  python local_test.py
> [1,0]<stderr>:Traceback (most recent call last):
> [1,0]<stderr>:  File "local_test.py", line 8, in <module>
> [1,0]<stderr>:    c = ht.concatenate((a, b), axis=-1)
> [1,0]<stderr>:  File "/Users/c.comito/HAF/heat/heat/core/manipulations.py", line 108, in concatenate
> [1,0]<stderr>:    ' {}, {}'.format(arr0.gshape, arr1.gshape))
> [1,0]<stderr>:ValueError: Arrays cannot be concatenated, gshapes must be the same in every axis except the selected axis: (2, 4), (2, 7)
> [1,1]<stderr>:Traceback (most recent call last):
> [1,1]<stderr>:  File "local_test.py", line 8, in <module>
> [1,1]<stderr>:    c = ht.concatenate((a, b), axis=-1)
> [1,1]<stderr>:  File "/Users/c.comito/HAF/heat/heat/core/manipulations.py", line 108, in concatenate
> [1,1]<stderr>:    ' {}, {}'.format(arr0.gshape, arr1.gshape))
> [1,1]<stderr>:ValueError: Arrays cannot be concatenated, gshapes must be the same in every axis except the selected axis: (2, 4), (2, 7)

Reply:
> Hi Daniel @coquelin77,
> 
> I ran a few tests and as far as I can tell it works fine (except for the negative axis case).
> 
> I see that the concatenated tensor inherits the split value from the input tensors and is distributed, it makes sense but it's not what e.g. maximum() and minimum() are doing (and probably other reduction operations as well, I've got to check), so if we agree that's what we want, I'll make the relevant changes.

The reason that I did this was to minimize communication. I feel that it makes intuitive sense to inherent the split axis in this case. I do not handle the separate split case here as it would require more overhead. instead i leave it for the user to decide if they want to resplit the data. and if so then they can do it themselves.